### ISSUE TITLE
Memory leak when we interrupt the PDF loading

### ISSUE THREAD

**Issue by xandao-dev** [hrynko/vue-pdf-embed#160]
![image](https://github.com/hrynko/vue-pdf-embed/assets/22106326/7be61c0f-fc9a-4c16-ac58-a71c81b605c5)
![image](https://github.com/hrynko/vue-pdf-embed/assets/22106326/b5999028-7b4c-4630-96d6-fd6dedf2d5e5)

# How to reproduce

1. Go to the page that has the PDF and before it gets loaded exit and check the detached elements.

# Observations

Tools used to identify the Memory Leak: Edge Detached Elements.

Our application: A SPA and I this specific page we change the path by calling $router.push.

Our implementation:
<img width="217" alt="image" src="https://github.com/hrynko/vue-pdf-embed/assets/22106326/96587d11-4292-4fa9-bbfa-2a994ef36ee5">

**Comment by hrynko** [comment]
Hi @xandao-dev,

Thanks for discovering this issue. I will try to fix it, but I would also appreciate the help in resolving it.

**Comment by xandao-dev** [comment]
@hrynko hey!

I was able to reproduce the issue in your project by slightly modifying the demo app to: 
```js
<template>
  <div>
    <button @click="enabled = !enabled">
      {{ enabled ? 'Disable' : 'Enable' }}
    </button>
    <vue-pdf-embed
      v-if="enabled"
      :image-resources-path="annotationIconsPath"
      :source="pdfSource"
    />
  </div>
</template>

<script>
import VuePdfEmbed from '../src/vue-pdf-embed.vue'

export default {
  components: {
    VuePdfEmbed,
  },
  data() {
    return {
      enabled: true,
      annotationIconsPath: '/node_modules/pdfjs-dist/web/images/',
      pdfSource: '/demo/sample-50-MB-pdf-file.pdf',
    }
  },
}
</script>
```

Note the pdf source of 50MB, we have to add it to the demo folder. Here is the link for the one I used: https://www.learningcontainer.com/sample-pdf-files-for-testing/

To check the memory leak, just spam clicks on the Enable/Disable button, and watch the detached elements tool.

**Comment by xandao-dev** [comment]
Looking closer at the first image I sent, it seems to be a problem with the **pdf.GlobalWorkerOptions.workerPort**
<img width="436" alt="image" src="https://github.com/hrynko/vue-pdf-embed/assets/22106326/3af8c10e-cf27-4ec2-a7e0-68c60a4ef966">

By moving the creation of the pdf worker instance to the Vue created lifecycle and then correctly terminating the worker in the beforeDestroy lifecycle hook, we can ensure proper worker termination.

I also read this issue about properly closing the loading task, but I haven't tested: https://github.com/mozilla/pdf.js/issues/11595

**Comment by xandao-dev** [comment]
@hrynko Can you please check if there are any drawbacks to loading the instance during the created lifecycle?

**Comment by hrynko** [comment]
Hi @xandao-dev,

Thank you for opening the PR, it looks good to me. I don't see any drawbacks in moving the worker initialization, however could you comment on the motivation for doing so?

**Comment by xandao-dev** [comment]
Sure, the worker does not get initialized again after termination in the beforeDestroy/beforeUnmount hook. Therefore, when the component is loaded again, it needs to be started again.

**Comment by xandao-dev** [comment]
If the worker is heavy, maybe it's an overkill to have an instance for each component. We do have alternatives, for example:

Stick with only one "global" (not really global, it belongs to a closure) worker, and in the beforeDestroy hook, we destroy the loading tasks instead of the worker.

Like so: loadingTask.destroy();

The option I followed is more memory-safe I guess, but it's also a bit more heavy (I don't know how much heavier).

**Comment by xandao-dev** [comment]
@hrynko hold on a bit, I would like to investigate a bit more with multiple components.

**Comment by xandao-dev** [comment]
The workers are not getting removed (in that first solution should exist one per pdf):

<img width="1139" alt="image" src="https://github.com/hrynko/vue-pdf-embed/assets/22106326/4fdc42fc-33d0-4a75-b519-51406dec865b">

I was dumb, GlobalWorkerOption has Global in its name for a reason. If you would like to use a worker for each instance you have to update [other parts of the code](https://github.com/mozilla/pdf.js/issues/16429). I tried to experiment with it to see if it handles better parallel PDF loading, but I wasn't able to accomplish it.

---
So let's take a step back, I analyzed the memory leak again I saw that was something related to event listeners.
I also concluded that `this.document?.destroy()` that already exists in the destruction hooks is enough to clean up the loading tasks and the worker.

So just by cleaning these event listeners the memory leak has gone!
<img width="353" alt="image" src="https://github.com/hrynko/vue-pdf-embed/assets/22106326/945e7ca7-3822-4b64-8dcb-1b4d1dc4bda3">

I'll be updating the PR.

**Comment by xandao-dev** [comment]
Here is the template I used to test the memory leak and the issues with multiple pdfs:

```js
<template>
  <div class="container">
    <div class="minw">
      <button @click="enabled = !enabled">
        {{ enabled ? 'Disable 1' : 'Enable 1' }}
      </button>
      <vue-pdf-embed
        v-if="enabled"
        :image-resources-path="annotationIconsPath"
        :source="pdfSource"
      />
    </div>

    <div class="minw">
      <button @click="enabled2 = !enabled2">
        {{ enabled2 ? 'Disable 2' : 'Enable 2' }}
      </button>
      <vue-pdf-embed
        v-if="enabled2"
        :image-resources-path="annotationIconsPath"
        :source="pdfSource"
      />
    </div>

    <div class="minw">
      <button @click="enabled3 = !enabled3">
        {{ enabled2 ? 'Disable 3' : 'Enable 3' }}
      </button>
      <vue-pdf-embed
        v-if="enabled3"
        :image-resources-path="annotationIconsPath"
        :source="pdfSource"
      />
    </div>

    <button
      @click="
        enabled3 = !enabled3
        enabled2 = !enabled2
        enabled = !enabled
      "
    >
      Toggle All
    </button>
  </div>
</template>

<script>
import VuePdfEmbed from '../src/vue-pdf-embed.vue'

export default {
  components: {
    VuePdfEmbed,
  },
  data() {
    return {
      enabled: true,
      enabled2: true,
      enabled3: true,
      annotationIconsPath: '/node_modules/pdfjs-dist/web/images/',
      pdfSource: '/demo/sample-50-MB-pdf-file.pdf',
    }
  },
}
</script>

<style lang="scss">
body {
  padding: 16px;
  background-color: #ccc;
}

.container {
  display: flex;
  justify-content: center;
  gap: 16px;
  max-width: 480px;
  margin: auto;
}

.minw {
  min-width: 300px;
}

.vue-pdf-embed {
  margin: auto;
  max-width: 480px;

  & > div {
    margin-bottom: 4px;
    box-shadow: 0 2px 8px 4px rgba(0, 0, 0, 0.1);
  }
}
</style>
```

**Comment by hrynko** [comment]
> Some memory leaks were caused by event listeners, by cleaning them it was fixed.

So does this fix the initial issue? If so, could you please update the PR?

**Comment by xandao-dev** [comment]
@hrynko all done, PR updated.

**Comment by xandao-dev** [comment]
@hrynko It's fine here. Should I've seen anything in the console?

<img width="1876" alt="image" src="https://github.com/hrynko/vue-pdf-embed/assets/22106326/d11181bb-c698-4029-af85-35bc0e068ecb">
<img width="1426" alt="image" src="https://github.com/hrynko/vue-pdf-embed/assets/22106326/eca4263c-a84e-4645-baf9-656a937493a7">

ps. This detached element is from an extension and is not a memory leak.

**Comment by xandao-dev** [comment]
> Shouldn't `documentLoadingTask` be reset as well?

this.document?.destroy() is basically destroying the documentLoadingTask. We can see it by debugging:

<img width="298" alt="image" src="https://github.com/hrynko/vue-pdf-embed/assets/22106326/0aa21f64-8bd9-470a-b51f-303babee4aad"><br>
<img width="284" alt="image" src="https://github.com/hrynko/vue-pdf-embed/assets/22106326/0fe9777b-7de8-4f86-b204-9349ccb5770f"><br>
<img width="622" alt="image" src="https://github.com/hrynko/vue-pdf-embed/assets/22106326/c1f93ffb-0449-4b8b-9514-b5697250eb93">

**Comment by hrynko** [comment]
I used your example with the Enable/Disable button and provided a large document. Once I click the "Disable" button (while loading the document), it throws the following error in the `render` method (as it was before the changes were made):

<img width="415" alt="image" src="https://github.com/hrynko/vue-pdf-embed/assets/28267443/ca81863a-f46e-4577-8c3e-8d0161c940c7">

**Comment by xandao-dev** [comment]
Hmmm, can you share the pdf, please? This error is not happening here.
It's even better if you could share a stackblitz or codesandbox with the issue happening.

**Comment by hrynko** [comment]
> @hrynko hey!
> 
> I was able to reproduce the issue in your project by slightly modifying the demo app to:
> 
> ```js
> <template>
>   <div>
>     <button @click="enabled = !enabled">
>       {{ enabled ? 'Disable' : 'Enable' }}
>     </button>
>     <vue-pdf-embed
>       v-if="enabled"
>       :image-resources-path="annotationIconsPath"
>       :source="pdfSource"
>     />
>   </div>
> </template>
> 
> <script>
> import VuePdfEmbed from '../src/vue-pdf-embed.vue'
> 
> export default {
>   components: {
>     VuePdfEmbed,
>   },
>   data() {
>     return {
>       enabled: true,
>       annotationIconsPath: '/node_modules/pdfjs-dist/web/images/',
>       pdfSource: '/demo/sample-50-MB-pdf-file.pdf',
>     }
>   },
> }
> </script>
> ```
> 
> Note the pdf source of 50MB, we have to add it to the demo folder. Here is the link for the one I used: https://www.learningcontainer.com/sample-pdf-files-for-testing/
> 
> To check the memory leak, just spam clicks on the Enable/Disable button, and watch the detached elements tool.

I applied the latest changes from the PR and the demo code as above. The PDF used can be found by the following link: https://history.nasa.gov/alsj/a17/A17_FlightPlan.pdf

The only thing I added is the logging of the `rendering-failed` event. It happens when I click "Disable" for the first time, while the doc is loading.

**Comment by xandao-dev** [comment]
Ahhh, now I see, is just the component event: 
<img width="823" alt="image" src="https://github.com/hrynko/vue-pdf-embed/assets/22106326/2af82c2a-6096-48e7-8ecc-7ef44a984aa4">

> The issue seems to be back

Well, that's not related to the initial issue and it's also not related to the memory leak. The behavior seems correct to me, it's an exception raised when the loading is interrupted.

**Comment by xandao-dev** [comment]
To make it clear to see what I solved, try to reproduce it with the **detached elements tool** available in the Edge browser.
You can also use the three-snapshot technique, but it's easier with this tool.

<img width="1175" alt="image" src="https://github.com/hrynko/vue-pdf-embed/assets/22106326/88c400be-3160-4362-bb78-91b4f58d7fc2">

Spam the button for a while, hit to get detached elements, hit the garbage collection, and get detached elements again. Finally, click to analyze the memory.

A detached element is an element that has been removed from the DOM tree but is still referenced by JavaScript code. As a result, it cannot be garbage-collected, leading to a memory leak (not always). They can prevent an entire tree of components from being garbage collected with a simple small leak in an inner component.

**Comment by hrynko** [comment]
Ok, it's clear now. Is there anything else you would like to change/add in this PR?

**Comment by xandao-dev** [comment]
@hrynko we are good to go, thanks.

**Comment by xandao-dev** [comment]
@hrynko Hey, don't want to bother you, but when can I expect a release?

**Comment by hrynko** [comment]
Hi @xandao-dev,

It has already been released in `v1.2.0`. Thanks again for your efforts, your input is always welcome!
